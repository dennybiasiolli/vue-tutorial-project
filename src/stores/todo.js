import { defineStore } from 'pinia'

export const defaultState = {
  items: [],
}

export const getters = {
  todoItems() {
    return this.items.filter(item => !item.completed)
  },
  completedItems() {
    return this.items.filter(item => item.completed)
  }
}

export const actions = {
  switchTodo(id, completed) {
    const item = this.items.find(item => item.id === id)
    item.completed = completed;
  },
  addTodo(text) {
    const maxId = Math.max(...this.items.map(item => item.id))
    this.items.push({
      id: maxId + 1,
      text,
      completed: false,
    })
  }
}

export const useTodoStore = defineStore('todo', {
  state: () => defaultState,
  getters,
  actions,
})
