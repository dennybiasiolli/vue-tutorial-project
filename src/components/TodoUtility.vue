<script setup>
import { storeToRefs } from 'pinia';
import { useTodoStore } from '@/stores/todo';
import { ref, onMounted, watch } from 'vue'
import TodoList from './TodoList.vue'

const todoStore = useTodoStore()

const formValid = ref(false)
const itemText = ref('')
const { todoItems, completedItems } = storeToRefs(todoStore)

const { getTodoItems, switchTodo, updateTodo, addTodo, deleteTodo } = todoStore

onMounted(() => {
  getTodoItems()
})

function handleSwitchTodo({ id, completed }) {
  switchTodo(id, completed)
}

function validateForm() {
  if (!itemText.value.trim()) {
    formValid.value = false
    return
  }
  formValid.value = true
}

function handleSubmit() {
  addTodo(itemText.value)
  itemText.value = ''
}

watch([itemText], () => {
  validateForm()
})
</script>

<template>
  <h2>Todo Utility</h2>

  <form @submit="handleSubmit">
    <input type="text" placeholder="To-do item" v-model="itemText" />
    <button type="submit" :disabled="!formValid">Add</button>
  </form>

  <TodoList
    v-if="todoItems.length > 0"
    title="To-do list"
    :items="todoItems"
    @switch-completed-todo="handleSwitchTodo"
    @change-todo-text="updateTodo($event.id, $event.text)"
    @delete-todo-item="deleteTodo"
  />

  <TodoList
    v-if="completedItems.length > 0"
    title="Completed list"
    :items="completedItems"
    @switch-completed-todo="switchTodo($event.id, $event.completed)"
    @change-todo-text="updateTodo($event.id, $event.text)"
    @delete-todo-item="deleteTodo"
  />
</template>

<style>
.item-completed {
  color: grey;
}
</style>
