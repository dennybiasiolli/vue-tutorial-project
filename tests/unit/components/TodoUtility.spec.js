import { createTestingPinia } from '@pinia/testing'
import { shallowMount } from '@vue/test-utils'
import { expect, it, beforeEach, vi } from 'vitest'
import TodoUtility from '@/components/TodoUtility.vue'
import { useTodoStore } from '@/stores/todo'

let pinia, todoStore

const myShallowMount = () => shallowMount(TodoUtility, {
  global: { plugins: [pinia] }
})

beforeEach(() => {
  pinia = createTestingPinia({
    createSpy: vi.fn,
    initialState: {
      todo: {
        items: [
          { id: 1, text: 'Todo item 1', completed: false },
          { id: 2, text: 'Todo item 2', completed: true },
        ],
      },
    },
  })
  todoStore = useTodoStore()
})

it('renders as expected', () => {
  const wrapper = myShallowMount()
  expect(wrapper.html()).toMatchSnapshot()
})

it('should switch todo', async () => {
  const wrapper = myShallowMount()
  const todoLists = wrapper.findAll('todo-list-stub')
  expect(todoLists.length).toBe(2)
  await todoLists[0].trigger('switch-completed-todo', { id: 1, completed: true })
  expect(todoStore.switchTodo).toHaveBeenCalledWith(1, true)
})
