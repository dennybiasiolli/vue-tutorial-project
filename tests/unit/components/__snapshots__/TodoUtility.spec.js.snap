// Vitest Snapshot v1

exports[`renders as expected 1`] = `
"<h2>Todo Utility</h2>
<form><input type=\\"text\\" placeholder=\\"To-do item\\"><button type=\\"submit\\" disabled=\\"\\">Add</button></form>
<todo-list-stub title=\\"To-do list\\" items=\\"[object Object]\\"></todo-list-stub>
<todo-list-stub title=\\"Completed list\\" items=\\"[object Object]\\"></todo-list-stub>"
`;
